awscli/topictags.py

```diff
@@ -0,0 +1,254 @@
+# Copyright (c) 2015 Amazon.com, Inc. or its affiliates.  All Rights Reserved
+#
+# Permission is hereby granted, free of charge, to any person obtaining a
+# copy of this software and associated documentation files (the
+# "Software"), to deal in the Software without restriction, including
+# without limitation the rights to use, copy, modify, merge, publish, dis-
+# tribute, sublicense, and/or sell copies of the Software, and to permit
+# persons to whom the Software is furnished to do so, subject to the fol-
+# lowing conditions:
+#
+# The above copyright notice and this permission notice shall be included
+# in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+# OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABIL-
+# ITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+# SHALL THE AUTHOR BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+# WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#
+import os
+import json
+import docutils.core
+
+
+class TopicTagDB(object):
+    """This class acts like a database for the tags of all available topics.
+
+    A tag is an element in a topic reStructured text file that contains
+    information about a topic. Information can range from titles to even
+    related CLI commands. Here are all of the currently supported tags:
+
+    Tag                 Meaning                         Required?
+    ---                 -------                         ---------
+    :title:             The title of the topic          Yes
+    :description:       Sentence description of topic   Yes
+    :category:          Category topic falls under      Yes
+    :related topic:     A related topic                 No
+    :related command:   A related command               No
+
+    To see examples of how to specify tags looks in the directory
+    awscli/topics. Note that tags can have multiple values by delimiting
+    values with commas. All tags must be on their own line in the file.
+
+    This class can load a JSON index represeting all of topics and their tags,
+    scan all of the topics and store the values of their tags, retrieve the
+    tag value for a particular topic, query for all the topics with a specific
+    tag and/or value, and save the loaded data back out to a JSON index.
+
+    The structure of the database can be viewed as a python dictionary:
+
+    {'topic-name-1': {
+        'title': ['My First Topic Title'],
```

Comment:
The only reason I kept it as a list is that it simplifies the code since you can always expect a list, and I would not have to do any non-list handling. Having multiple elements in the tile or description list would be looked at as a bug or issue in the source file. Whenever, I try to grab the description or title (from outside the class), I will always try to grab the first element in that list, which should be the only element.

I agree the semantics is a bit weird. I just choose it to keep it that way to help simplify/shorten the codebase. @jamesls Let me know what you think.